### PR TITLE
feat(eth-wire): RLP encode then compress

### PR DESCRIPTION
**Pull request by Rjected**
Previously the `Encodable` and `Decodable` implementations for `p2p` messages like `Disconnect` were brittle and included logic related to snappy compression. This is incorrect, because the `Encodable` and `Decodable` implementations should only define what is necessary to encode and decode the RLP representation of the data type, and should not include logic related to protocol details like snappy.

Now, we assume the input to the `Decodable` implementations are not compressed, and we do not encode compressed messages with `Encodable` implementations.

This effects serialization for:
 - `DisconnectReason`
 - `P2PMessage::Ping`
 - `P2PMessage::Pong`
 
 `start_disconnect` is refactored to return a `Result<(), snap::Error>`, and it now compresses the disconnect message explicitly.

Tests are removed which ensure `Encodable` implementations are snappy decodable.
Tests are removed which ensure `Decodable` implementations work correctly with snappy-compressed input.

Documentation is added which explains the need for capability offsets, how they are calculated, and how they are used to determine subprotocol message IDs.

### CODE REVIEW

File: crates/net/eth-wire/src/p2pstream.rs [hunk 1]
```diff
@@ -199,14 +199,43 @@ impl<S> P2PStream<S> {
...
+        let mut compressed = BytesMut::zeroed(1 + snap::raw::max_compress_len(buf.len() - 1));
+        let compressed_size = self.encoder.compress(&buf[1..], &mut compressed[1..])?;
+
+        // truncate the compressed buffer to the actual compressed size (plus one for the message
+        // id)
+        compressed.truncate(compressed_size + 1);
+
+        // we do not add the capability offset because the disconnect message is a `p2p` reserved
+        // message
+        compressed[0] = buf[0];
```
Comment (gakonst) [inline crates/net/eth-wire/src/p2pstream.rs:228]: Should this go to a function and be unit tested?

File: crates/net/eth-wire/src/p2pstream.rs [hunk 2]
```diff
@@ -199,14 +199,43 @@ impl<S> P2PStream<S> {
...
+    pub fn start_disconnect(&mut self, reason: DisconnectReason) -> Result<(), snap::Error> {
         // clear any buffered messages and queue in
         self.outgoing_messages.clear();
         let disconnect = P2PMessage::Disconnect(reason);
         let mut buf = BytesMut::with_capacity(disconnect.length());
         disconnect.encode(&mut buf);
-        self.outgoing_messages.push_back(buf.freeze());
+
+        tracing::trace!(
+            fromlen=%buf.len(),
```
Comment (gakonst) [inline crates/net/eth-wire/src/p2pstream.rs:214]: len_before

File: crates/net/eth-wire/src/p2pstream.rs [hunk 3]
```diff
@@ -199,14 +199,43 @@ impl<S> P2PStream<S> {
...
+        // truncate the compressed buffer to the actual compressed size (plus one for the message
+        // id)
+        compressed.truncate(compressed_size + 1);
+
+        // we do not add the capability offset because the disconnect message is a `p2p` reserved
+        // message
+        compressed[0] = buf[0];
+
+        tracing::trace!(
+            tolen=%compressed.len(),
```
Comment (gakonst) [inline crates/net/eth-wire/src/p2pstream.rs:231]: len_after

**Review by gakonst — APPROVED** [review]
Need to rebase and remove Matt's fork of disc5

**Comment by gakonst** [comment]
Just repeating what you did:
* before: encodable did both rlp + snappy. this was error prone and was not clean because we did snappy inside the encodable/decodable traits. This also required extra tests there to ensure the mixed encoding with compression was correct
* after: we pull snappy in the messaging functions operating on the rlp encoded buffers directly, instead of doing it inside the type. this allows us to remove some tests and reduce complexity 

Is that about right? Anything else to be mindful of?

**Review by mattsse — APPROVED** [review]
great, 
this way, it's way easier to understand where snappy is applied.

**Comment by gakonst** [comment]
Merged optimistically

**Pull request merged**